src/js/program/program-controller.js

```diff
@@ -44,58 +71,64 @@ class ProgramController extends Events {
-            if (casting || this._providerCanPlay(mediaController.provider, source)) {
-                // We can synchronously reuse the current mediaController
-                this.loadPromise = Promise.resolve(mediaController);
-                // Reinitialize the mediaController with the new item, allowing a new playback session
-                mediaController.activeItem = item;
-                this._setActiveMedia(mediaController);
-                return this.loadPromise;
+            model.setActiveItem(index);
+
+            const source = getSource(item);
+            if (!source) {
+                return Promise.reject(new PlayerError(MSG_CANT_PLAY_VIDEO, ERROR_PLAYLIST_ITEM_MISSING_SOURCE));
             }
 
-            // If we can't play the source with the current provider, reset the current one and
-            // prime the next tag within the gesture
-            this._destroyActiveMedia();
-        }
+            // Activate the background media if it's loading the item we want to play
+            if (background.isNext(playlistItem)) {
+                // First destroy the active item so that the BGL provider can enter the foreground
+                this._destroyActiveMedia();
+                // Attach the BGL provider into the load/play chain
+                return this._activateBackgroundMedia();
+            }
 
-        const mediaModelContext = model.mediaModel;
-        this.loadPromise = this._setupMediaController(source)
-            .then(nextMediaController => {
+            if (mediaController) {
+                const casting = model.get('castActive');
+                if (casting || this._providerCanPlay(mediaController.provider, source)) {
+                    // We can synchronously reuse the current mediaController
+                    // Reinitialize the mediaController with the new item, allowing a new playback session
+                    mediaController.activeItem = playlistItem;
+                    this._setActiveMedia(mediaController);
+                    return mediaController;
+                }
+
+                // If we can't play the source with the current provider, reset the current one and
+                // prime the next tag within the gesture
+                this._destroyActiveMedia();
+            }
+
+            const mediaModelContext = model.mediaModel;
+            return this._setupMediaController(source).then(nextMediaController => {
                 // Don't do anything if we've tried to load another provider while this promise was resolving
                 // We check using the mediaModel because it is unique per item, and per instance of that item
                 if (mediaModelContext === model.mediaModel) {
                     nextMediaController.activeItem = item;
                     this._setActiveMedia(nextMediaController);
                     return nextMediaController;
                 }
-            })
-            .catch(err => {
+            }).catch(err => {
```

Comment:
this is my lack of understanding of bg loading alongside some difficult to parse changes. can we maybe walk through this file in person or if you feel like it's explainable by text feel free to just reply

Reply:
Let's talk through it. It wasn't easy for me either. This code and it's execution is complex and I don't intend to commit it to memory.

Reply:
scheduled some time for tuesday